torch/autograd/gradcheck.py

```diff
@@ -725,47 +716,65 @@ def _differentiable_outputs(x):
     return tuple(o for o in _as_tuple(x) if o.requires_grad)
 
 
-def get_notallclose_msg(analytical, numerical, output_idx, input_idx, error_str='') -> str:
-    return error_str + 'Jacobian mismatch for output %d with respect to input %d,\n' \
+def get_notallclose_msg(analytical, numerical, output_idx, input_idx, complex_indices, inp_is_complex) -> str:
+    out_is_imag = complex_indices and output_idx in complex_indices
+    grad_output = "1j" if out_is_imag else "1"
+    prefix = "" if not inp_is_complex and not out_is_imag else \
+        f"Gradients failed to compare equal for grad output = {grad_output}"
+    return prefix + 'Jacobian mismatch for output %d with respect to input %d,\n' \
         'numerical:%s\nanalytical:%s\n' % (output_idx, input_idx, numerical, analytical)
 
+
 def transpose(matrix_of_tensors):
     # returns list of tuples
     return list(zip(*matrix_of_tensors))
 
 
+def real_and_imag(fn, sample_outputs):
+    # returns new functions real(fn), and imag(fn) where real(fn) and imag(fn) behave the same as
+    # the original fn, except torch.real or torch.imag are applied to the complex outputs
+    def apply_to_c_outs(fn, fn_to_apply):
+        def wrapped_fn(*inputs):
+            outs = _as_tuple(fn(*inputs))
+            return tuple(fn_to_apply(o) if o.is_complex() else o for o in outs)
+        return wrapped_fn
+    return apply_to_c_outs(fn, torch.real), apply_to_c_outs(fn, torch.imag)
+
+
+def gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps, rtol,
+                        atol, check_grad_dtypes, nondet_tol, complex_indices, any_outputs_complex):
+    if any_outputs_complex:
+        real_fn, imag_fn = real_and_imag(func, outputs)
+
+        imag_func_out = imag_fn(*tupled_inputs)
+        imag_outputs = _differentiable_outputs(imag_func_out)
+        gradcheck_fn(imag_fn, imag_func_out, tupled_inputs, imag_outputs, eps,
+                     rtol, atol, check_grad_dtypes, nondet_tol, complex_indices)
+
+        real_func_out = real_fn(*tupled_inputs)
+        real_outputs = _differentiable_outputs(real_func_out)
+        gradcheck_fn(real_fn, real_func_out, tupled_inputs, real_outputs, eps,
+                     rtol, atol, check_grad_dtypes, nondet_tol)
```

Comment:
You're missing complex_indices no?

Reply:
Good catch!

Reply:
This isn't being tested though, which is why we did not catch this. We'd need a function that fails while considering the real part of a complex output, but succeeds when considering only the imaginary part, or maybe there is some way to accomplish this with hooks.

Reply:
Yes I think you can add some hooks on the result and add noise to either the real or imaginary part of the gradients?